File: tools/flow/tokenizer/portable.py
import typing

from .base import Token, EOT
from .data import FlowData

__all__ = [
    'make_tokenizer_function'
]


def make_tokenizer_function(managers: FlowData) -> typing.Callable[[str], typing.Iterator[Token]]:
    def tokenizer(text: str) -> typing.Iterator[Token]:
        state = 0
        content = ''
        at = 0
        to = 0
        for char in text + EOT:
            while char:
                handlers, action = managers[state]
                for chars, action in handlers:
                    if char in chars:
                        break
                if not action:
                    raise NotImplementedError
                options, build = action[0]
                if options & 1:  # add
                    content += char
                if options & 2:  # inc
                    to += 1
                if options & 4:  # clr
                    char = None
                if build:  # build
                    token = Token(type=build, content=content, at=at, to=to)
                    yield token
                if options & 8:  # clear
                    content = ''
                    at = to
                state = action[1]
    
    return tokenizer

File: tools/flow/tokenizer/portable.js
export default function makeTokenizer(managers) {
  return function tokenizer(text) {
    let state = 0;
    let content = '';
    let at = 0;
    let to = 0;
    let tokens = [];
    for (let char of text + '\0') {
      while (char !== null) {
        const manager = managers[state];
        let action = manager[1];
        for (const handler of manager[0]) {
          if (handler[0].includes(char)) {
            action = handler[1];
            break;
          }
        }
        if (!action) {
          throw new SyntaxError();
        }
        const params = action[0];
        if (params[0]) // add
          content += char;
        if (params[1]) // use
          to++
        if (params[2]) // clr
          char = null;
        if (params[3]) { // build
          let token = {type: params[3], content: content, at: at, to: to};
          tokens.push(token)
        }
        if (params[4]) { // clear
          content = '';
          at = to;
        }
        state = action[1];
      }
    }
    return tokens;
  }
}

